text/0092-contract-source-registry.md

```diff
@@ -0,0 +1,159 @@
+
+## General
+
+* Contract sources are indexed by code hash and not by contract address. This means that several deployed instances of the same smart contract code will only need to be registered once.
+
+* If a smart contract supports code updates, once its code is changed, the new code hash will no longer be found in the registry until a source code matching the new code hash is uploaded.
+
+* IPFS is used as a **temporary** form of storage only until TON Storage is released. All participants commit to performing a full migration of all verified sources to TON Storage when possible.
+
+* A single participating **verifier** runs a backend that exposes HTTP API that receives a list of source files, compiles them and returns a signed proof over the resulting code hash. The public key for this verifier is registered in the verifier registry contract.
+
+* To encourage redundancy of verifiers, each verifier is encouraged to run multiple backend instances, each with its own public key. Verification then requires a quorum of backends to work and not all of them, for example 3 out of 5. This is optional and a verifier can choose to run a single backend with a single public key too. This redundancy allows for easy participation of decentralized oracle networks like Orbs Network in the protocol (networks that have dozens of independent validators) as a single verifier.
+
+## sources.json
+
+JSON file provided by a specific **verifier** for a specific **contract code hash** containing the URLs of source-code files and verification attestations. Fields of this file include:
+
+* `codeHash` - SHA256 of the code cell CellRepr (https://ton-blockchain.github.io/docs/tvm.pdf 3.1.5), e.g. E/XXoxbG124QU+iKxZtd5loHKjiEUTcdxcW+y7oT9Q4=
+* `sources` - Array of sources files, each with the following fields:
+    * `url` - Human-readable source-code URL on IPFS and later TON Storage, e.g. [ipfs://QmWQE1HhYuWieFcocfZZvK4gvJqgswNpj3zUCvdPNCC12T](https://tonsource.infura-ipfs.io/ipfs/QmWQE1HhYuWieFcocfZZvK4gvJqgswNpj3zUCvdPNCC12T)
+    * `originalFilename` - e.g. "wallet-v3.fc" or "imports/stdlib.fc"
+* `compilerType` - Compiler used to compile the verified sources, e.g. "func"
+* `compilerVersion` - Version of compiler, e.g. "0.2.0"
+* `commandLine` - Command-line used to compile sources, e.g. "func -o output.fif -SPA wallet-v3.fc stdlib.fc"
+* `verificationTime` - Unix timestamp of when verification was made
+
+## Verifier registry contract
+
+A smart contract deployed to TON mainnet that holds a mapping between a **verifier id** to the **verifier details** which include the list of backends, their public keys and quorum configuration. To prevent spam in this registry, we propose that each verifier will deposit in the contract a sum of 1,000-10,000 TON coin. This sum will be returned when the verifier unregisters.
```

Comment:
An interesting point was whether to tie the staking amount to a [config param](https://github.com/ton-blockchain/ton/blob/35d17249e6b54d67a5781ebf26e4ee98e56c1e50/crypto/block/block.tlb#L648), in order to reflect changes in TON/USD price.

E.g. 10 * 1e9 * Gas prices (which amounts to 10K ton)

Reply:
It's a good idea, since the contract is immutable without any special admin role, this can normalize the deposit size in case TON USD price changes significantly

Reply:
> An interesting point was whether to tie the staking amount to a [config param](https://github.com/ton-blockchain/ton/blob/35d17249e6b54d67a5781ebf26e4ee98e56c1e50/crypto/block/block.tlb#L648), in order to reflect changes in TON/USD price.
> 
> 
> 
> E.g. 10 * 1e9 * Gas price (currently 1,000) => 10K ton
> 
> 
> 
>

Reply:
> > An interesting point was whether to tie the staking amount to a [config param](https://github.com/ton-blockchain/ton/blob/35d17249e6b54d67a5781ebf26e4ee98e56c1e50/crypto/block/block.tlb#L648), in order to reflect changes in TON/USD price.
> > 
> > 
> > 
> > E.g. 10 * 1e9 * Gas price (currently 1,000) => 10K ton
> > 
> > 
> > 
> > 
> 
>

Reply:
> It's a good idea, since the contract is immutable without any special admin role, this can normalize the deposit size in case TON USD price changes significantly

Reply:
> > It's a good idea, since the contract is immutable without any special admin role, this can normalize the deposit size in case TON USD price changes significantly
> 
>

Reply:
Cool